File: dotnet/Cucumber.Query/Query.cs
#nullable enable

using Cucumber.Messages;
using Io.Cucumber.Messages.Types;
using System;
using System.Collections.Generic;
using System.Linq;

namespace Cucumber.Query;

public class Query
{
    private readonly Repository _repository;

    public Query(Repository repository)
    {
        _repository = repository;
    }

    public IDictionary<TestStepResultStatus, long> CountMostSevereTestStepResultStatus()
    {
        var statusCounts = new Dictionary<TestStepResultStatus, long>();
        // Initialize with zero for each possible TestStepResultStatus
        foreach (TestStepResultStatus status in Enum.GetValues(typeof(TestStepResultStatus)))
        {
            statusCounts[status] = 0;
        }
        foreach (var testCaseStarted in FindAllTestCaseStarted())
        {
            var finishedSteps = FindTestStepsFinishedBy(testCaseStarted).ToList();
            if (finishedSteps.Count == 0)
                continue;
            // Find the most severe status (largest enum value)
            var mostSevere = finishedSteps
                .Select(f => f.TestStepResult.Status)
                .Max();
            if (statusCounts.ContainsKey(mostSevere))
                statusCounts[mostSevere]++;
            else
                statusCounts[mostSevere] = 1;
        }
        return statusCounts;
    }

    public int TestCasesStartedCount => FindAllTestCaseStarted().Count();

    public IEnumerable<Pickle> FindAllPickles() => _repository.PickleById.Values;

    public IEnumerable<PickleStep> FindAllPickleSteps() => _repository.PickleStepById.Values;

    public IEnumerable<TestCaseStarted> FindAllTestCaseStarted() => _repository.TestCaseStartedById.Values
        .Where(tcs => !FindTestCaseFinishedBy(tcs)?.WillBeRetried ?? true); // Exclude finished cases that will be retried

    public IEnumerable<StepDefinition> FindAllStepDefinitions() => _repository.StepDefinitionById.Values;

    public IEnumerable<TestCaseFinished> FindAllTestCaseFinished()
    {
        return _repository.TestCaseFinishedByTestCaseStartedId.Values
            .Where(tcFinished => !tcFinished.WillBeRetried);
    }

    public IEnumerable<TestStep> FindAllTestSteps() => _repository.TestStepById.Values;

    public IEnumerable<TestCase> FindAllTestCases()
    {
        return _repository.TestCaseById.Values;
    }

    public IEnumerable<TestStepStarted> FindAllTestStepStarted()
    {
        return _repository.TestStepsStartedByTestCaseStartedId.Values
            .SelectMany(list => list);
    }

    public IEnumerable<TestStepFinished> FindAllTestStepFinished()
    {
        return _repository.TestStepsFinishedByTestCaseStartedId.Values
            .SelectMany(list => list);
    }

    public IEnumerable<TestRunHookStarted> FindAllTestRunHookStarted()
    {
        return _repository.TestRunHookStartedById.Values;
    }

    public IEnumerable<TestRunHookFinished> FindAllTestRunHookFinished()
    {
        return _repository.TestRunHookFinishedByTestRunHookStartedId.Values;
    }

    public IEnumerable<UndefinedParameterType> FindAllUndefinedParameterTypes()
    {
        return _repository.UndefinedParameterTypes;
    }

    public IEnumerable<Attachment> FindAttachmentsBy(TestStepFinished testStepFinished) =>
        _repository.AttachmentsByTestCaseStartedId.TryGetValue(testStepFinished.TestCaseStartedId, out var attachments)
            ? attachments.Where(a => a.TestStepId == testStepFinished.TestStepId)
            : new List<Attachment>();

    public IEnumerable<Attachment> FindAttachmentsBy(TestRunHookFinished testRunHookFinished)
    {
        if (_repository.AttachmentsByTestRunHookStartedId.TryGetValue(testRunHookFinished.TestRunHookStartedId, out var attachments))
            return new List<Attachment>(attachments);
        return new List<Attachment>();
    }

    public Hook? FindHookBy(TestStep testStep)
    {
        if (!string.IsNullOrEmpty(testStep.HookId) && _repository.HookById.TryGetValue(testStep.HookId, out var hook))
        {
            return hook;
        }
        return null;
    }

    public Hook? FindHookBy(TestRunHookStarted testRunHookStarted)
    {
        if (!string.IsNullOrEmpty(testRunHookStarted.HookId) && _repository.HookById.TryGetValue(testRunHookStarted.HookId, out var hook))
        {
            return hook;
        }
        return null;
    }

    public Hook? FindHookBy(TestRunHookFinished testRunHookFinished)
    {
        var testRunHookStarted = FindTestRunHookStartedBy(testRunHookFinished);
        if (testRunHookStarted == null)
        {
            return null;
        }
        return FindHookBy(testRunHookStarted);
    }

    public Meta? FindMeta() => _repository.Meta;

    public TestStepResult? FindMostSevereTestStepResultBy(TestCaseStarted testCaseStarted)
    {
        var finishedSteps = FindTestStepsFinishedBy(testCaseStarted);
        if (finishedSteps.Count() == 0)
            return null;
        // Find the TestStepFinished with the most severe status (highest enum value)
        var mostSevere = finishedSteps
            .OrderBy(f => f.TestStepResult.Status)
            .LastOrDefault();
        return mostSevere?.TestStepResult;
    }

    public TestStepResult? FindMostSevereTestStepResultBy(TestCaseFinished testCaseFinished)
    {
        var testCaseStarted = FindTestCaseStartedBy(testCaseFinished);
        return testCaseStarted != null ? FindMostSevereTestStepResultBy(testCaseStarted) : null;
    }

    public Location? FindLocationOf(Pickle pickle)
    {
        var lineage = FindLineageBy(pickle);
        if (lineage == null)
            return null;
        if (lineage.Example != null)
            return lineage.Example.Location;
        if (lineage.Scenario != null)
            return lineage.Scenario.Location;
        return null;
    }

    public Pickle? FindPickleBy(TestCaseStarted testCaseStarted)
    {
        var testCase = FindTestCaseBy(testCaseStarted);
        if (testCase != null && _repository.PickleById.TryGetValue(testCase.PickleId, out var pickle))
        {
            return pickle;
        }
        return null;
    }

    public Pickle? FindPickleBy(TestCaseFinished testCaseFinished)
    {
        var testCaseStarted = FindTestCaseStartedBy(testCaseFinished);
        return testCaseStarted != null ? FindPickleBy(testCaseStarted) : null;
    }

    public Pickle? FindPickleBy(TestCase testCase)
    {
        return testCase != null && _repository.PickleById.TryGetValue(testCase.PickleId, out var pickle) ? pickle : null;
    }

    public Pickle? FindPickleBy(TestStepStarted testStepStarted)
    {
        var testCaseStarted = FindTestCaseStartedBy(testStepStarted);
        return testCaseStarted != null ? FindPickleBy(testCaseStarted) : null;
    }

    public Pickle? FindPickleBy(TestStepFinished testStepFinished)
    {
        var testCaseStarted = FindTestCaseStartedBy(testStepFinished);
        return testCaseStarted != null ? FindPickleBy(testCaseStarted) : null;
    }

    public PickleStep? FindPickleStepBy(TestStep testStep)
    {
        if (!string.IsNullOrEmpty(testStep.PickleStepId))
        {
            if (_repository.PickleStepById.TryGetValue(testStep.PickleStepId, out var pickleStep))
            {
                return pickleStep;
            }
        }
        return null;
    }

    public IEnumerable<Suggestion> FindSuggestionsBy(PickleStep pickleStep)
    {
        if (_repository.SuggestionsByPickleStepId.TryGetValue(pickleStep.Id, out var suggestions))
            return new List<Suggestion>(suggestions);
        return new List<Suggestion>();
    }

    public IEnumerable<Suggestion> FindSuggestionsBy(Pickle pickle)
    {
        var result = new List<Suggestion>();
        foreach (var step in pickle.Steps)
            result.AddRange(FindSuggestionsBy(step));
        return result;
    }

    public Step? FindStepBy(PickleStep pickleStep)
    {
        if (pickleStep.AstNodeIds != null && pickleStep.AstNodeIds.Count > 0)
        {
            var stepId = pickleStep.AstNodeIds[0];
            if (!string.IsNullOrEmpty(stepId) && _repository.StepById.TryGetValue(stepId, out var step))
            {
                return step;
            }
        }
        return null;
    }

    public IEnumerable<StepDefinition> FindStepDefinitionsBy(TestStep testStep)
    {
        if (testStep.StepDefinitionIds != null)
            return testStep.StepDefinitionIds
                .Select(id => _repository.StepDefinitionById.TryGetValue(id, out var def) ? def : null)
                .Where(def => def != null)
                .Cast<StepDefinition>();
        return new List<StepDefinition>();
    }

    public StepDefinition? FindUnambiguousStepDefinitionBy(TestStep testStep)
    {
        if (testStep.StepDefinitionIds != null && testStep.StepDefinitionIds.Count == 1)
            return _repository.StepDefinitionById.TryGetValue(testStep.StepDefinitionIds[0], out var def) ? def : null;
        return null;
    }

    public TestCase? FindTestCaseBy(TestCaseStarted testCaseStarted)
    {
        if (_repository.TestCaseById.TryGetValue(testCaseStarted.TestCaseId, out var testCase))
        {
            return testCase;
        }
        return null;
    }

    public TestCase? FindTestCaseBy(TestCaseFinished testCaseFinished)
    {
        var testCaseStarted = FindTestCaseStartedBy(testCaseFinished);
        return testCaseStarted != null ? FindTestCaseBy(testCaseStarted) : null;
    }

    public TestCase? FindTestCaseBy(TestStepStarted testStepStarted)
    {
        var testCaseStarted = FindTestCaseStartedBy(testStepStarted);
        return testCaseStarted != null ? FindTestCaseBy(testCaseStarted) : null;
    }

    public TestCase? FindTestCaseBy(TestStepFinished testStepFinished)
    {
        var testCaseStarted = FindTestCaseStartedBy(testStepFinished);
        return testCaseStarted != null ? FindTestCaseBy(testCaseStarted) : null;
    }

    public System.TimeSpan? FindTestCaseDurationBy(TestCaseStarted testCaseStarted)
    {
        var started = testCaseStarted.Timestamp;
        var finished = FindTestCaseFinishedBy(testCaseStarted)?.Timestamp;
        if (finished != null)
        {
            var startTime = Converters.ToDateTime(started);
            var finishTime = Converters.ToDateTime(finished);
            return finishTime - startTime;
        }
        return null;
    }

    public System.TimeSpan? FindTestCaseDurationBy(TestCaseFinished testCaseFinished)
    {
        var testCaseStarted = FindTestCaseStartedBy(testCaseFinished);
        return testCaseStarted != null ? FindTestCaseDurationBy(testCaseStarted) : null;
    }

    public TestCaseStarted? FindTestCaseStartedBy(TestStepStarted testStepStarted)
    {
        return _repository.TestCaseStartedById.TryGetValue(testStepStarted.TestCaseStartedId, out var tcs) ? tcs : null;
    }

    public TestCaseStarted? FindTestCaseStartedBy(TestCaseFinished testCaseFinished)
    {
        return _repository.TestCaseStartedById.TryGetValue(testCaseFinished.TestCaseStartedId, out var tcs) ? tcs : null;
    }

    public TestCaseStarted? FindTestCaseStartedBy(TestStepFinished testStepFinished)
    {
        return _repository.TestCaseStartedById.TryGetValue(testStepFinished.TestCaseStartedId, out var tcs) ? tcs : null;
    }

    public TestCaseFinished? FindTestCaseFinishedBy(TestCaseStarted testCaseStarted)
    {
        return _repository.TestCaseFinishedByTestCaseStartedId.TryGetValue(testCaseStarted.Id, out var finished) ? finished : null;
    }

    public TestRunHookFinished? FindTestRunHookFinishedBy(TestRunHookStarted testRunHookStarted)
    {
        return _repository.TestRunHookFinishedByTestRunHookStartedId.TryGetValue(testRunHookStarted.Id, out var finished) ? finished : null;
    }

    public TestRunHookStarted? FindTestRunHookStartedBy(TestRunHookFinished testRunHookFinished)
    {
        return _repository.TestRunHookStartedById.TryGetValue(testRunHookFinished.TestRunHookStartedId, out var started) ? started : null;
    }

    public System.TimeSpan? FindTestRunDuration()
    {
        if (_repository.TestRunStarted == null || _repository.TestRunFinished == null)
            return null;
        var start = Converters.ToDateTime(_repository.TestRunStarted.Timestamp);
        var finish = Converters.ToDateTime(_repository.TestRunFinished.Timestamp);
        return finish - start;
    }

    public TestRunFinished? FindTestRunFinished() => _repository.TestRunFinished;

    public TestRunStarted? FindTestRunStarted() => _repository.TestRunStarted;

    public TestStep? FindTestStepBy(TestStepStarted testStepStarted)
    {
        return _repository.TestStepById.TryGetValue(testStepStarted.TestStepId, out var testStep) ? testStep : null;
    }

    public TestStep? FindTestStepBy(TestStepFinished testStepFinished)
    {
        return _repository.TestStepById.TryGetValue(testStepFinished.TestStepId, out var testStep) ? testStep : null;
    }

    public IEnumerable<TestStepStarted> FindTestStepsStartedBy(TestCaseStarted testCaseStarted)
    {
        if (_repository.TestStepsStartedByTestCaseStartedId.TryGetValue(testCaseStarted.Id, out var steps))
        {
            return new List<TestStepStarted>(steps);
        }
        return new List<TestStepStarted>();
    }

    public IEnumerable<TestStepStarted> FindTestStepsStartedBy(TestCaseFinished testCaseFinished)
    {
        if (_repository.TestStepsStartedByTestCaseStartedId.TryGetValue(testCaseFinished.TestCaseStartedId, out var steps))
        {
            return new List<TestStepStarted>(steps);
        }
        return new List<TestStepStarted>();
    }

    public IEnumerable<TestStepFinished> FindTestStepsFinishedBy(TestCaseStarted testCaseStarted)
    {
        if (_repository.TestStepsFinishedByTestCaseStartedId.TryGetValue(testCaseStarted.Id, out var steps))
        {
            return new List<TestStepFinished>(steps);
        }
        return new List<TestStepFinished>();
    }

    public IEnumerable<TestStepFinished> FindTestStepsFinishedBy(TestCaseFinished testCaseFinished)
    {
        var testCaseStarted = FindTestCaseStartedBy(testCaseFinished);
        return testCaseStarted != null ? FindTestStepsFinishedBy(testCaseStarted) : new List<TestStepFinished>();
    }

    public IEnumerable<(TestStepFinished, TestStep)> FindTestStepFinishedAndTestStepBy(TestCaseStarted testCaseStarted)
    {
        var finishedSteps = FindTestStepsFinishedBy(testCaseStarted);
        var result = new List<(TestStepFinished, TestStep)>();
        foreach (var testStepFinished in finishedSteps)
        {
            var testStep = FindTestStepBy(testStepFinished);
            if (testStep != null)
            {
                result.Add((testStepFinished, testStep));
            }
        }
        return result;
    }

    // FindLineageBy methods
    public Lineage? FindLineageBy(GherkinDocument element)
    {
        _repository.LineageById.TryGetValue(element, out var lineage);
        return lineage;
    }

    public Lineage? FindLineageBy(Feature element)
    {
        _repository.LineageById.TryGetValue(element, out var lineage);
        return lineage;
    }

    public Lineage? FindLineageBy(Rule element)
    {
        _repository.LineageById.TryGetValue(element, out var lineage);
        return lineage;
    }

    public Lineage? FindLineageBy(Scenario element)
    {
        _repository.LineageById.TryGetValue(element, out var lineage);
        return lineage;
    }

    public Lineage? FindLineageBy(Examples element)
    {
        _repository.LineageById.TryGetValue(element, out var lineage);
        return lineage;
    }

    public Lineage? FindLineageBy(TableRow element)
    {
        _repository.LineageById.TryGetValue(element, out var lineage);
        return lineage;
    }

    public Lineage? FindLineageBy(Pickle pickle)
    {
        var astNodeIds = pickle.AstNodeIds;
        var lastAstNodeId = astNodeIds.LastOrDefault();
        _repository.LineageById.TryGetValue(lastAstNodeId, out var lineage);
        return lineage;
    }

    public Lineage? FindLineageBy(TestCaseStarted testCaseStarted)
    {
        var pickle = FindPickleBy(testCaseStarted);
        if (pickle == null)
        {
            return null;
        }
        return FindLineageBy(pickle);
    }

    public Lineage? FindLineageBy(TestCaseFinished testCaseFinished)
    {
        var pickle = FindPickleBy(testCaseFinished);
        if (pickle == null)
        {
            return null;
        }
        return FindLineageBy(pickle);
    }
}

File: ruby/lib/cucumber/query.rb
# frozen_string_literal: true

require 'cucumber/messages'

require_relative 'repository'

# Given one Cucumber Message, find another.
#
# Queries can be made while the test run is incomplete - and this will naturally return incomplete results
# see {Cucumber Messages - Message Overview}[https://github.com/cucumber/messages?tab=readme-ov-file#message-overview]
#
module Cucumber
  # Provides lookup methods for related Cucumber messages stored in the `Repository` class.
  class Query
    attr_reader :repository
    private :repository

    include Cucumber::Messages::Helpers::TimeConversion
    include Cucumber::Messages::Helpers::TestStepResultComparator

    def initialize(repository)
      @repository = repository
    end

    # @return [Integer]
    def count_test_cases_started
      find_all_test_case_started.length
    end

    # @return [Array<Pickle>]
    def find_all_pickles
      repository.pickle_by_id.values
    end

    # @return [Array<PickleStep>]
    def find_all_pickle_steps
      repository.pickle_step_by_id.values
    end

    # @return [Array<StepDefinition>]
    def find_all_step_definitions
      repository.step_definition_by_id.values
    end

    # @return [Array<TestCaseStarted>]
    # This finds all test cases from the following conditions (UNION)
    #   -> Test cases that have started, but not yet finished
    #   -> Test cases that have started, finished, but that will NOT be retried
    def find_all_test_case_started
      repository.test_case_started_by_id.values.select do |test_case_started|
        test_case_finished = find_test_case_finished_by(test_case_started)
        test_case_finished.nil? || !test_case_finished.will_be_retried
      end
    end

    # @return [Array<TestCaseFinished>]
    # This finds all test cases that have finished AND will not be retried
    def find_all_test_case_finished
      repository.test_case_finished_by_test_case_started_id.values.reject(&:will_be_retried)
    end

    # @return [Array<TestCase>]
    def find_all_test_cases
      repository.test_case_by_id.values
    end

    # @return [Array<TestRunHookStarted>]
    def find_all_test_run_hook_started
      repository.test_run_hook_started_by_id.values
    end

    # @return [Array<TestRunHookFinished>]
    def find_all_test_run_hook_finished
      repository.test_run_hook_finished_by_test_run_hook_started_id.values
    end

    # @return [Array<TestStepStarted>]
    def find_all_test_step_started
      repository.test_steps_started_by_test_case_started_id.values.flatten
    end

    # @return [Array<TestStepFinished>]
    def find_all_test_step_finished
      repository.test_steps_finished_by_test_case_started_id.values.flatten
    end

    # @return [Array<TestSteps>]
    def find_all_test_steps
      repository.test_step_by_id.values
    end

    # @param (message) [TestStep, TestRunHookStarted, TestRunHookFinished]
    # @return [Array<Hook>, nil]
    def find_hook_by(message)
      ensure_only_message_types!(
        message,
        [Cucumber::Messages::TestStep, Cucumber::Messages::TestRunHookStarted, Cucumber::Messages::TestRunHookFinished],
        '#find_hook_by'
      )

      if message.is_a?(Cucumber::Messages::TestRunHookFinished)
        test_run_hook_started_message = find_test_run_hook_started_by(message)
        test_run_hook_started_message && find_hook_by(test_run_hook_started_message)
      else
        repository.hook_by_id[message.hook_id]
      end
    end

    # @return [Meta]
    def find_meta
      repository.meta
    end

    # @param (message) [TestCaseStarted, TestCaseFinished]
    # @return [String, nil]
    def find_most_severe_test_step_result_by(message)
      ensure_only_message_types!(
        message,
        [Cucumber::Messages::TestCaseStarted, Cucumber::Messages::TestCaseFinished],
        '#find_most_severe_test_step_result_by'
      )

      if message.is_a?(Cucumber::Messages::TestCaseStarted)
        find_test_steps_finished_by(message)
          .map(&:test_step_result)
          .max_by { |test_step_result| test_step_result_rankings[test_step_result.status] }
      else
        test_case_started_message = find_test_case_started_by(message)
        test_case_started_message && find_most_severe_test_step_result_by(test_case_started_message)
      end
    end

    # @param (message) [TestCase, TestCaseStarted, TestCaseFinished, TestStepStarted, TestStepFinished]
    # @return [Pickle]
    def find_pickle_by(message)
      ensure_only_message_types!(
        message,
        [
          Cucumber::Messages::TestCase,
          Cucumber::Messages::TestCaseStarted,
          Cucumber::Messages::TestCaseFinished,
          Cucumber::Messages::TestStepStarted,
          Cucumber::Messages::TestStepFinished
        ],
        '#find_pickle_by'
      )

      test_case_message = message.is_a?(Cucumber::Messages::TestCase) ? message : find_test_case_by(message)
      repository.pickle_by_id[test_case_message.pickle_id]
    end

    # @param (message) [TestStep]
    # @return [PickleStep]
    def find_pickle_step_by(message)
      ensure_only_message_types!(
        message,
        [Cucumber::Messages::TestStep],
        '#find_pickle_step_by'
      )

      repository.pickle_step_by_id[message.pickle_step_id]
    end

    # @param (message) [PickleStep]
    # @return [Step]
    def find_step_by(message)
      ensure_only_message_types!(
        message,
        [Cucumber::Messages::PickleStep],
        '#find_step_by'
      )

      repository.step_by_id[message.ast_node_ids.first]
    end

    # @param (message) [TestStep]
    # @return [Array<StepDefinition>]
    def find_step_definitions_by(message)
      ensure_only_message_types!(
        message,
        [Cucumber::Messages::TestStep],
        '#find_step_definitions_by'
      )

      ids = message.step_definition_ids.nil? ? [] : message.step_definition_ids
      ids.filter_map { |id| repository.step_definition_by_id[id] }
    end

    # @param (message) [TestCaseStarted, TestCaseFinished, TestStepStarted, TestStepFinished]
    # @return [TestCase]
    def find_test_case_by(message)
      ensure_only_message_types!(
        message,
        [
          Cucumber::Messages::TestCaseStarted,
          Cucumber::Messages::TestCaseFinished,
          Cucumber::Messages::TestStepStarted,
          Cucumber::Messages::TestStepFinished
        ],
        '#find_test_case_by'
      )

      test_case_started_message = message.is_a?(Cucumber::Messages::TestCaseStarted) ? message : find_test_case_started_by(message)
      repository.test_case_by_id[test_case_started_message.test_case_id]
    end

    # @param (message) [TestCaseFinished, TestStepStarted, TestStepFinished]
    # @return [TestCaseStarted]
    def find_test_case_started_by(message)
      ensure_only_message_types!(
        message,
        [Cucumber::Messages::TestCaseFinished, Cucumber::Messages::TestStepStarted, Cucumber::Messages::TestStepFinished],
        '#find_test_case_started_by'
      )

      repository.test_case_started_by_id[message.test_case_started_id]
    end

    # @param (message) [TestCaseStarted]
    # @return [TestCaseFinished]
    def find_test_case_finished_by(test_case_started)
      ensure_only_message_types!(
        test_case_started,
        [Cucumber::Messages::TestCaseStarted],
        '#find_test_case_finished_by'
      )

      repository.test_case_finished_by_test_case_started_id[test_case_started.id]
    end

    # @return [Hash<String, Integer>, nil]
    def find_test_run_duration
      if repository.test_run_started.nil? || repository.test_run_finished.nil?
        nil
      else
        float_time = timestamp_to_time(repository.test_run_finished.timestamp) - timestamp_to_time(repository.test_run_started.timestamp)
        seconds_to_duration(float_time)
      end
    end

    # @param (message) [TestRunHookFinished]
    # @return [TestRunHookStarted]
    def find_test_run_hook_started_by(message)
      ensure_only_message_types!(
        message,
        [Cucumber::Messages::TestRunHookFinished],
        '#find_test_run_hook_started_by'
      )

      repository.test_run_hook_started_by_id[message.test_run_hook_started_id]
    end

    # @param (message) [TestRunHookStarted]
    # @return [TestRunHookFinished]
    def find_test_run_hook_finished_by(message)
      ensure_only_message_types!(
        message,
        [Cucumber::Messages::TestRunHookStarted],
        '#find_test_run_hook_finished_by'
      )

      repository.test_run_hook_finished_by_test_run_hook_started_id[message.id]
    end

    # @return [TestRunStarted]
    def find_test_run_started
      repository.test_run_started
    end

    # @return [TestRunFinished]
    def find_test_run_finished
      repository.test_run_finished
    end

    # @param (message) [TestStepStarted, TestStepFinished]
    # @return [TestStep]
    def find_test_step_by(message)
      ensure_only_message_types!(
        message,
        [Cucumber::Messages::TestStepStarted, Cucumber::Messages::TestStepFinished],
        '#find_test_step_by'
      )

      repository.test_step_by_id[message.test_step_id]
    end

    # @param (message) [TestCaseStarted, TestCaseFinished]
    # @return [Array<TestStep>]
    def find_test_steps_started_by(message)
      ensure_only_message_types!(
        message,
        [Cucumber::Messages::TestCaseStarted, Cucumber::Messages::TestCaseFinished],
        '#find_test_steps_started_by'
      )

      key = message.is_a?(Cucumber::Messages::TestCaseStarted) ? message.id : message.test_case_started_id
      repository.test_steps_started_by_test_case_started_id.fetch(key, [])
    end

    # @param (message) [TestCaseStarted, TestCaseFinished]
    # @return [Array<TestStep>]
    def find_test_steps_finished_by(message)
      ensure_only_message_types!(
        message,
        [Cucumber::Messages::TestCaseStarted, Cucumber::Messages::TestCaseFinished],
        '#find_test_steps_finished_by'
      )

      if message.is_a?(Cucumber::Messages::TestCaseStarted)
        repository.test_steps_finished_by_test_case_started_id.fetch(message.id, [])
      else
        test_case_started_message = find_test_case_started_by(message)
        test_case_started_message.nil? ? [] : find_test_steps_finished_by(test_case_started_message)
      end
    end

    private

    def ensure_only_message_types!(supplied_message, permissible_message_types, method_name)
      raise ArgumentError, "Supplied argument is not a Cucumber Message. Argument: #{supplied_message.class}" unless supplied_message.is_a?(Cucumber::Messages::Message)
      raise ArgumentError, "Supplied message type '#{supplied_message.class}' is not permitted to be used when calling #{method_name}" unless permissible_message_types.include?(supplied_message.class)
    end
  end
end
